baybe/simulation/lookup.py

```diff
@@ -0,0 +1,209 @@
+"""Target lookup mechanisms."""
+
+import warnings
+from typing import Callable, Collection, List, Literal, Union
+
+import numpy as np
+import pandas as pd
+
+from baybe.targets import NumericalTarget, TargetMode
+from baybe.targets.base import Target
+from baybe.utils import add_fake_results
+
+CallableLookup = Callable[[pd.DataFrame], pd.DataFrame]
+"""A callable that can be used to retrieve target values for given parameter
+configurations."""
+
+
+def look_up_targets(
+    queries: pd.DataFrame,
+    targets: List[Target],
+    lookup: Union[pd.DataFrame, CallableLookup, None] = None,
+    impute_mode: Literal["error", "worst", "best", "mean", "random"] = "error",
+) -> None:
+    """Fill the target values for given parameter configurations via a lookup mechanism.
+
+    Note that this does not create a new dataframe but modifies ``queries`` in-place.
+
+    Args:
+        queries: A dataframe containing the parameter configurations.
+        targets: The targets whose values should be filled.
+        lookup: An optional lookup mechanism, provided in the form of a dataframe
+            or callable, that defines the targets for the queried parameter settings.
+            If omitted, fake values will be used.
+
+            - pd.DataFrame: Each row contains the one parameter configuration and the
+              corresponding target values.
+            - callable: A callable that can be used to retrieve target values for given
+              parameter configurations. Expects a dataframe where each row corresponds
+              to one parameter configuration (columns names being the parameter names)
+              and must return a dataframe containing the corresponding target values
+              (column names being the target names), where each row holds the targets
+              for the corresponding parameter row.
+        impute_mode: Specifies how a missing lookup will be handled (only relevant
+            for dataframe lookups). See :func:`baybe.simulation.core.simulate_campaign`
+            for details.
+
+    Raises:
+        AssertionError: If an analytical function is used and an incorrect number of
```

Comment:
Shouldn't this be an "or" instead of an "and"?

Reply:
No, in the original context it should "and" – note that I did not modify the message. But that's why it's even better that you pointed it out, because I forgot to rewrite it during the refactor. Neither the Error type nor the message is no longer up to date :D Have changed it now, thanks 👏🏼 Please resolve if OK.

Reply:
The function can also raise an `AssertionError` via `_lookup_targets_from_dataframe`, so this should also be added here.